### DISCUSSION TITLE
Safe mode/encourage making schema changes on the lowest environment on the Cloud

### DISCUSSION THREAD

**Discussion by Jay-umbr** [umbraco/Umbraco.Cloud.Issues#785] [Ideas]
With the way the Cloud is set up - where schema between environments is synchronized by Git deployments - it is possible to mess up your project quite a bit by making changes directly on Live, and then deploying from Dev to Live (if you happen to have more than one environment). It would be nice if the users would have the option to enable/disable a "safe mode" - and in that case, let's say:

-you have Dev and Live environments
-you have safe mode enabled
-you log into Dev backoffice, can make changes in doctypes/templates/datatypes etc. just fine
-you log into Live backoffice, but whenever you try to make a change/save changes in schema on Live, a popup would appear saying something like

"Safe mode is enabled. You cannot save any schema changes to prevent possible conflicts between environments. If you wish to modify the schema, please make the changes on the Development Environment and deploy them all the way up.
To disable safe mode please go to" ... etc etc

With this setting in place, and enabled, it would enforce the good practice of making changes on Dev instead of Live, if you have more than one environment. That could prevent schema mismatch errors, and colliding datatypes/templates/etc,

**Comment by nul800sebastiaan** [comment]
Interesting, that would probably require us to significantly change how Umbraco works. 🙈

**Comment by sitereactor** [comment]
It's something we've talked about before in relation to vNext and having the Live environment somewhere else. If the Settings section could be read-only it would better support the flow of changes (typically done by a developer in the dev-environment or local) are pushed forward without the risk of something having been changed in another environment.

While we can remove the sections to enforce this it would be a lot nicer if certain sections could be set to read-only.

**Comment by nul800sebastiaan** [comment]
This will go on our long-term ideas list then for now! 👍

**Comment by skttl** [comment]
That's a nice idea. I think it can be done with event handlers on the different save events if you need it now :)

**Comment by thomasknielsen** [comment]
In the same lines - it might be worth considering to make it possible to disable the automatic patch updates of Umbraco if you have a local dev environment often it would be nice to control what version of Umbraco is used and plan the upgrade to push out from local->dev->live.

**Comment by sitereactor** [comment]
@skttl yes it could be done by cancelling events, but I would prefer a friendly'er approach where its obvious that what you are looking at is read-only. Maybe greyed out icons and editor, maybe a banner of sorts and no save button.

@thomasknielsen I think that is a bit of a stretch to say its related 😉 but scheduling updates is something that is in the cards. It's still early though, so difficult to say how it would work, but I imagine some kind of schedule per Project, which the Project Owner can control, so updates are either applied right away or on a controlled schedule.

**Comment by robmaas** [comment]
I have described a similar FR on Umbraco-CMS. It's not a Cloud-only thing, however the implementation will probably be a bit different for Cloud projects.
https://github.com/umbraco/Umbraco-CMS/issues/8862

**Comment by matthewcare** [comment]
This is something I'd like to see also.
My current approach for restricting access is with a rewrite rule such as
```
<rule name="Block Umbraco Edit Screens">
  <match url="*" />
  <conditions logicalGrouping="MatchAny">
    <add input="{URL}" pattern="/umbraco/views/documentTypes/edit.html" />
     .....
  </conditions>
  <action type="Rewrite" url="/restricted.html"  />
</rule>
```
so it would be nice to have a proper way to disable it.

**Comment by ronaldbarendse** [comment]
Thanks for the idea. Six years on it is still relevant, but this is something the CMS needs to implement first, so I am closing this discussion in favour of the CMS repository.

The reason is that a read-only schema mode is only useful if the backoffice shows it as read-only (greyed-out editors, no save button, a notice explaining where to make the change) and the Management API enforces it for anything that bypasses the UI. Cloud or Deploy could only cancel save notifications after the fact, which gives editors an error instead of guidance. That is the same conclusion reached earlier in this thread, and the related CMS request (umbraco/Umbraco-CMS#8862) was never picked up.

The good news is that the CMS team now has an open proposal for exactly this: umbraco/Umbraco-CMS#23289 makes document types, media types, member types, data types, templates, partial views, scripts and stylesheets read-only in the `Production` runtime mode, both in the backoffice and in the Management API. It is a proposal without a timeline yet, so please treat it as a direction rather than a promise. The principle behind it is that a fully pre-compiled release package should not change any schema. That is not a Cloud-only need: any artifact-based deployment, whether an immutable build, a container image or infrastructure as code, treats the deployed artifact as the source of truth for schema, and schema edits on the deployed environment are drift.

This matters for Cloud and Deploy in particular. On Cloud your schema lives as UDA files in the environment's repository and Deploy applies them to the database of each environment through the CMS service layer. Editing schema directly on Live writes new UDA files there that do not exist in the lower environments, which is where the schema mismatches and conflicting data types you mention come from. With `ModelsMode` set to `Nothing`, those edits are also never reflected in the compiled models. The proposal therefore blocks the backoffice and the Management API but keeps the service layer writable, so Deploy can still align each environment with the deployed UDA files while nothing else can change schema that would require regenerating those files and the models.

You can get part of this today, and we recommend it for every Cloud Live environment: run Live in the `Production` runtime mode (https://docs.umbraco.com/umbraco-cms/run-in-production/runtime-modes). With `Umbraco:CMS:Runtime:Mode` set to `Production`:

- Templates and partial views can't be edited in the backoffice, because runtime compilation is disabled and views are compiled at build/publish time;
- The Models Builder dashboard is disabled, because `ModelsMode` must be `Nothing`, so document type changes made on Live would not end up in your models anyway.

The Cloud Portal already requires Production mode before load balancing can be enabled (https://docs.umbraco.com/umbraco-cloud/optimize-and-maintain-your-site/optimize-performance/load-balancing), so it is a supported setup on Cloud. Do read the prerequisites first: Production mode validates a Release build, precompiled views, `ModelsMode: Nothing`, `UseHttps` and `UmbracoApplicationUrl` at boot, and fails to start if they are not met. Until the read-only mode lands, the supported way to keep editors out of schema on Live is user group permissions that hide the Settings section on that environment.

If you want to influence the outcome, the proposal has open questions on whether Languages, Webhooks and Document Blueprints should be in scope, and whether dictionary keys should be locked while translations stay editable. Please comment on the PR with your view.
